File: api/src/routes/ec.ts
import { Router, Request, Response } from "express";
import prisma from "../lib/prisma.js";
import { z } from "zod";
import { runScoutAgent } from "../agents/scoutAgent.js";

const router = Router();

const getStudentId = async (userId: number, role: string, targetStudentId?: string) => {
	const userRole = role.toUpperCase();
	if (userRole === "STUDENT") {
		const student = await prisma.student.findUnique({ where: { userId } });
		return student?.id;
	}
	if ((userRole === "ADMIN" || userRole === "COUNSELOR") && targetStudentId) {
		return parseInt(targetStudentId);
	}
	return null;
};

router.get("/", async (req: Request, res: Response) => {
	const user = res.locals.user;
	if (!user) return res.status(401).json({ error: "Unauthorized" });

	const studentId = await getStudentId(user.id, user.role, req.query.studentId as string);
	console.log(`[EC GET] User: ${user.email}, Role: ${user.role}, Requested StudentId: ${req.query.studentId}, Resolved StudentId: ${studentId}`);

	if (!studentId) return res.status(400).json({ error: "Student ID required or not found" });

	const ecs = await prisma.extracurricular.findMany({
		where: { studentId },
		orderBy: { createdAt: "desc" }
	});

	console.log(`[EC GET] Found ${ecs.length} records`);
	res.json(ecs);
});

router.post("/find-clubs", async (req: Request, res: Response) => {
	const user = res.locals.user;
	if (!user) return res.status(401).json({ error: "Unauthorized" });

	const studentId = await getStudentId(user.id, user.role, req.body.studentId);
	if (!studentId) return res.status(400).json({ error: "Student ID required or not found" });

	const config = await prisma.systemConfig.findFirst();
	const agentConfig = config?.agentConfig as any;
	if (!agentConfig?.scout?.enabled) {
		return res.status(400).json({ error: "Opportunity Scout is disabled" });
	}

	// Run synchronously for now to remove Redis dependency
	runScoutAgent(studentId).catch(console.error);

	res.json({ message: "Scout agent triggered", studentId });
});

const ecSchema = z.object({
	name: z.string(),
	role: z.string().optional(),
	impactDescription: z.string().optional(),
	hoursPerWeek: z.number().optional(),
	weeksPerYear: z.number().optional(),
	proofUrl: z.string().optional(),
	studentId: z.number().optional()
});

router.post("/", async (req: Request, res: Response) => {
	const user = res.locals.user;
	if (!user) return res.status(401).json({ error: "Unauthorized" });

	const result = ecSchema.safeParse(req.body);
	if (!result.success) return res.status(400).json({ error: result.error.format() });

	let studentId = result.data.studentId;
	if (user.role.toUpperCase() === "STUDENT") {
		const student = await prisma.student.findUnique({ where: { userId: user.id } });
		if (!student) return res.status(404).json({ error: "Student record not found" });
		studentId = student.id;
	} else if (!studentId) {
		return res.status(400).json({ error: "studentId is required for non-students" });
	}

	const ec = await prisma.extracurricular.create({
		data: {
			...result.data,
			studentId: studentId!
		}
	});

	res.status(201).json(ec);
});

router.delete("/:id", async (req: Request, res: Response) => {
	const user = res.locals.user;
	if (!user) return res.status(401).json({ error: "Unauthorized" });

	const id = parseInt(req.params.id);
	const ec = await prisma.extracurricular.findUnique({ where: { id } });
	if (!ec) return res.status(404).json({ error: "Record not found" });

	if (user.role.toUpperCase() === "STUDENT") {
		const student = await prisma.student.findUnique({ where: { userId: user.id } });
		if (!student || ec.studentId !== student.id) {
			return res.status(403).json({ error: "Forbidden" });
		}
	}

	await prisma.extracurricular.delete({ where: { id } });
	res.status(204).end();
});

export default router;

File: api_py/routers/ec.py
from fastapi import APIRouter, Depends, HTTPException, Request
from ..database import execute_query, execute_commit
from ..auth import get_current_user
from pydantic import BaseModel
from typing import Optional
import traceback

router = APIRouter()

class ECRecord(BaseModel):
    name: str
    role: Optional[str] = None
    impactDescription: Optional[str] = None
    hoursPerWeek: Optional[int] = 0
    weeksPerYear: Optional[int] = 0
    studentId: Optional[int] = None

def get_target_student_id(user, requested_student_id):
    role = user["role"].upper()
    if role == "STUDENT":
        student = execute_query("SELECT id FROM Student WHERE userId = %s", (user["id"],), fetch_one=True)
        return student["id"] if student else None
    if (role in ["ADMIN", "COUNSELOR"]):
        if requested_student_id:
            return int(requested_student_id)
        return None
    return None

@router.get("/")
def get_ecs(studentId: Optional[str] = None, current_user: dict = Depends(get_current_user)):
    try:
        target_id = get_target_student_id(current_user, studentId)
        if not target_id:
            if current_user["role"].upper() in ["ADMIN", "COUNSELOR"]:
                return []
            raise HTTPException(status_code=400, detail="Student ID required")
            
        ecs = execute_query("SELECT * FROM Extracurricular WHERE studentId = %s ORDER BY createdAt DESC", (target_id,))
        return ecs
    except HTTPException:
        raise
    except Exception as e:
        print(f"[EC] Exception in GET /: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/")
def add_ec(ec: ECRecord, current_user: dict = Depends(get_current_user)):
    try:
        target_id = get_target_student_id(current_user, ec.studentId)
        if not target_id:
            raise HTTPException(status_code=400, detail="Student ID required")
            
        query = """
            INSERT INTO Extracurricular (studentId, name, role, impactDescription, hoursPerWeek, weeksPerYear)
            VALUES (%s, %s, %s, %s, %s, %s)
        """
        params = (target_id, ec.name, ec.role, ec.impactDescription, ec.hoursPerWeek, ec.weeksPerYear)
        ec_id = execute_commit(query, params)
        return {"id": ec_id, **ec.dict()}
    except HTTPException:
        raise
    except Exception as e:
        print(f"[EC] Exception in POST /: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))

@router.delete("/{id}/")
def delete_ec(id: int, current_user: dict = Depends(get_current_user)):
    try:
        user_id = current_user["id"]
        user_role = current_user["role"].upper()
        
        if user_role == "STUDENT":
            student = execute_query("SELECT id FROM Student WHERE userId = %s", (user_id,), fetch_one=True)
            if not student:
                raise HTTPException(status_code=404, detail="Student record not found")
            
            record = execute_query("SELECT studentId FROM Extracurricular WHERE id = %s", (id,), fetch_one=True)
            if not record or record["studentId"] != student["id"]:
                raise HTTPException(status_code=403, detail="Forbidden")
                
        execute_commit("DELETE FROM Extracurricular WHERE id = %s", (id,))
        return {"success": True}
    except HTTPException:
        raise
    except Exception as e:
        print(f"[EC] Exception in DELETE /{id}: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/find-clubs")
def find_clubs(studentId: int, current_user: dict = Depends(get_current_user)):
    return {"message": "Scout agent triggered (Simulated)", "studentId": studentId}
